### PR TITLE
kvstore: Introduce shared store type

### PR DESCRIPTION
**Pull request by tgraf**
Package store implements a shared store backed by a kvstore or similar with
the following properties:

* A single type is used to represent all keys
* Any number of collaborators can join the store. Typically a collaborator
  is an individual Cilium agent running on each node.
* All collaborators can own and contribute keys to the store. Each key is
  owned by exactly one collaborator. It is the responsibility of each
  collaborator to pick a key name which is guaranteed to be unique.
* All collaborate desire to see all keys within the scope of a store. The
  scope of the store is defined by a common key prefix. For this purpose,
  each collaborator maintains a local cache of all keys in the store by
  subscribing to change events.

This implementation will be used by the upcoming kvstore based node discovery

Signed-off-by: Thomas Graf <thomas@cilium.io>

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/4124)

### CODE REVIEW

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

File: pkg/kvstore/store/store.go [hunk 1]
```diff
@@ -0,0 +1,429 @@
...
+	synchronizationIntervalDefault = time.Minute
+
+	// watcherChanSize is the size of the channel to buffer kvstore events
+	watcherChanSize = 100
+)
+
+var (
+	controllers controller.Manager
+
+	log = logrus.WithField(logfields.LogSubsys, "shared-store")
```
Comment (ianvernon) [inline pkg/kvstore/store/store.go]: this should be `logging.DefaultLogger.WithField...` where logging is `pkg/logging`

File: pkg/kvstore/store/store.go [hunk 2]
```diff
@@ -0,0 +1,429 @@
...
+
+	if err := s.syncLocalKey(key); err != nil {
+		s.DeleteLocalKey(key)
+		return err
+	}
+
+	return nil
+}
+
+// DeleteLocalKey removes a key from bein synchronized with the kvstore
```
Comment (ianvernon) [inline pkg/kvstore/store/store.go]: bein --> being

File: pkg/kvstore/store/store.go [hunk 3]
```diff
@@ -0,0 +1,429 @@
...
+func (s *SharedStore) DeleteLocalKey(key LocalKey) {
+	err := kvstore.Delete(s.keyPath(key))
+	name := key.GetKeyName()
+
+	s.mutex.Lock()
+	delete(s.localKeys, name)
```
Comment (ianvernon) [inline pkg/kvstore/store/store.go]: if you `delete(s.localKeys, name)` and then check `_, ok := s.localKeys[name]`, won't `ok` always be `false`?

File: pkg/kvstore/store/store.go [hunk 4]
```diff
@@ -0,0 +1,429 @@
...
+				close(listDone)
+				continue
+			}
+
+			logger := s.getLogger().WithFields(logrus.Fields{
+				"key":       event.Key,
+				"eventType": event.Typ,
+			})
+
+			logger.Infof("Received key update via kvstore [value %s]", event.Value)
```
Comment (ianvernon) [inline pkg/kvstore/store/store.go]: Based on past experience, you may not have the value of the event when the event is `kvstore.EventTypeDelete`, so the log message might look strange.

File: pkg/kvstore/store/store.go [hunk 5]
```diff
@@ -0,0 +1,429 @@
...
+		"storeName": s.name,
+	})
+}
+
+func (s *SharedStore) updateKey(name string, value []byte) error {
+	s.mutex.Lock()
+	key := s.sharedKeys[name]
+
+	// shared key is seen for the first time
+	if key == nil {
```
Comment (ianvernon) [inline pkg/kvstore/store/store.go]: why not:

``` key, ok := s.sharedKeys[name]
if !ok { ... }
```

?

File: pkg/kvstore/store/store.go [hunk 6]
```diff
@@ -0,0 +1,429 @@
...
+		if s.localKeys[name] != nil {
```
Comment (ianvernon) [inline pkg/kvstore/store/store.go]: Same here, why not use the `key, ok` convention?

File: pkg/kvstore/store/store.go [hunk 7]
```diff
@@ -0,0 +1,429 @@
...
+	return nil
+}
+
+func (s *SharedStore) deleteKey(name string) {
+	s.mutex.Lock()
+	existingKey := s.sharedKeys[name]
+	delete(s.sharedKeys, name)
+	s.mutex.Unlock()
+
+	if existingKey == nil {
```
Comment (ianvernon) [inline pkg/kvstore/store/store.go]: And here.

**Review by ianvernon — CHANGES_REQUESTED** [review]
First round of comments.

**Comment by tgraf** [comment]
test-me-please

File: pkg/kvstore/store/store.go [hunk 8]
```diff
@@ -0,0 +1,429 @@
...
+				close(listDone)
+				continue
+			}
+
+			logger := s.getLogger().WithFields(logrus.Fields{
+				"key":       event.Key,
+				"eventType": event.Typ,
+			})
+
+			logger.Infof("Received key update via kvstore [value %s]", event.Value)
```
Comment (aanm) [inline pkg/kvstore/store/store.go]: The `event.Value` is a slice of byes so `event.Value` should be cast into a string

File: pkg/kvstore/store/store.go [hunk 9]
```diff
@@ -0,0 +1,429 @@
...
+
+			keyName := strings.TrimPrefix(event.Key, s.conf.Prefix)
+			if keyName[0] == '/' {
+				keyName = keyName[1:]
+			}
+
+			switch event.Typ {
+			case kvstore.EventTypeCreate, kvstore.EventTypeModify:
+				if err := s.updateKey(keyName, event.Value); err != nil {
+					logger.WithError(err).Warningf("Unable to unmarshal store value: %s", event.Value)
```
Comment (aanm) [inline pkg/kvstore/store/store.go]: The `event.Value` is a slice of byes so `event.Value` should be cast into a string

**Comment by tgraf** [comment]
test-me-please

**Comment by tgraf** [comment]
test-me-please

**Review by aanm — APPROVED** [review]

**Comment by tgraf** [comment]
test-me-please

**Pull request merged**